### DISCUSSION TITLE
App docker image version compatibility. How do we validate new images work with other services before deployment?

### DISCUSSION THREAD

**Discussion by Stramba** [BretFisher/ama#142] [Q&A]
Hi Bret!
We started using Docker and Swarm Mode recently: we grouped our services in multiple compose files, but we deploy a single stack with all of the services. 

We would like to know how to verify that APIs exposed by a given version of **serviceA** is compliant with what a given version of **serviceB** needs. There are tools (or best practice) that allow a version compatibility check before the deploy?

We know that integration test can solve this problem, but it makes sense to have something to ensure that stacks with non compliant images cannot be deployed (also in integration test environments)?

**Comment by BretFisher** [comment]
I'm not sure this is a docker thing. Basically what I'm hearing you ask is "how do I ensure my deployment toolchain doesn't mix incompatible versions when it deploys apps". That's something you'd need to test for.
So once you've done all CI testing, you'd have built images that you would push to your docker registry. Then you need to change something to deploy them... maybe it's Kubernetes manifest yaml, maybe it's something else, but that part, needs to be tested with functional testing across services. Only once those tests are passed, is the yaml change "approved" for prod deployment.
Does that make sense?

**Comment by BretFisher** [comment]
I changed the question to hopefully, more accurately describe your question. Feel free to edit.

**Comment by Stramba** [comment]
Thank you very much Bret! So, basically, we need functional tests in Swarm stack repository to verify whether that the stack correctly start and others checking API usage. Am I right? Do you have any advice about how to implement first kind of tests?

**Comment by BretFisher** [comment]
When you functional test something in CI, you should be testing it with the versions of other apps you plan to run it with. That testing should validate that your "new app version works with the other apps in the stack". This doesn't have to use swarm, just use docker-compose or however you test containers in your CI. This is something you were hopefully doing before containers. 

Then if you want an extra layer of testing, you would test your deployment yaml and automate the validation of your apps while running on "as close to production infrastructure as possible". Most teams don't get this advanced, even though it's very helpful in preventing issues during deployment.  Sadly, I don't know any tools to automate that for Swarm (you know you might be able to build a swarm inside a single docker container (docker-in-docker). Kubernetes does this too with the kind project https://kind.sigs.k8s.io/

**Comment by Stramba** [comment]
Ok, clear! The "extra layer" could be what we are looking for.

**Comment by BretFisher** [comment]
Not really. dind and kind is how most CI's do it.
